### DISCUSSION TITLE
Enhancement: Assets as extensions to system types

### DISCUSSION THREAD

**Discussion by KacperCzapp** [SwiftGen/SwiftGen#901] [Ideas]
I would like to be able to pass an optional parameter to the SwiftGen configuration so that assets are added as extensions to system types instead of / next to the usual SwiftGen structures. 

Examples:

### Colors
Color assets would be extensions to UIColor/NSColor:
```swift
// before:
view.backgroundColor = Assets.Colors.midnightBlue.color
// after:
view.backgroundColor = .midnightBlue
```

### Images
Image assets would be extensions to UIImage/NSImage:
```swift
// before:
imageView.image = Assets.WelcomeScreen.background
// after:
imageView.image = .welcomeScreen.background
```

### Strings:
```
"alert_title" = "Title of the alert";
"bananas.owner" = "Those %d bananas belong to %@.";
```

```swift
// before:
label1.text = L10n.alertTitle
label2.text = L10n.Bananas.owner(3, "John")
// after:
label1.text = .alertTitle
label2.text = .bananas.owner(3, "John")
```

Please let me know if my idea is valid and useful (by voting on the issue). This mechanism could also be applied to Fonts, maybe Storyboards. If time allows maybe I will try and implement it myself.

**Comment by BrentMifsud** [comment]
A problem with this proposal is that your assets can conflict with apples own built in assets.

swiftgen already generates its own initializers for most assets.

For example:

`UIColor(asset:)` and `UIImage(asset:)`

These are also not very convenient to use because they are optional. They are also more verbose than they need to be because you have to reference the main asset enum.

I propose that some new extensions are generated to solve this. Using Colors as an example:

Firstly, static ColorAssets should also be generated inside of the ColorAsset class.

```swift
extension ColorAsset {
    static let red = ColorAsset(name: "red")
    static let green = ColorAsset(name: "green")
    static let blue = ColorAsset(name: "blue")
}
```

Secondly a new function should be added to Color typealias: 

```swift
extension ColorAsset.Color {
    static func color(_ colorAsset: ColorAsset) -> Color {
        colorAsset.color
    }
}
```

The new usage would be:

```swift
view.backgroundColor = .color(.red)
```

Not as convenient as your proposal, but its less verbose than what we have today, and won't conflict with the built in colors/images.

**Reply by KacperCzapp** [reply]
> A problem with this proposal is that your assets can conflict with apples own built in assets.

Somehow, the compiler knows that if there is a user-defined color that has the same name as the system one, it chooses the user's one.

I have generated using my custom template the following code and I have no issues. 

```swift
#if os(iOS) || os(tvOS)
extension UIColor {
  internal static let black = UIColor(named: "black")!
}
```

Even defining it as `class var`, as in the system framework, still doesn't cause compiler issues, both at the definition and call site.

**Reply by contreras2004** [reply]
Hey @KacperCzapp ,  could you please share the template? I was looking for the same thing and I'm struggling to generate it
